### DISCUSSION TITLE
How to specify a specific package version to CIBW_TEST_REQUIRES for Windows

### DISCUSSION THREAD

**Discussion by bhazelton** [pypa/cibuildwheel#1749] [Q&A] [answered]
We've been using cibuildwheel very successfully for some time, it's wonderful! But a new problem just came up on windows builds because we need to limit pytest to <8.0 for now. The following is working on linux & mac builds but failing on windows:

```
CIBW_TEST_REQUIRES: packaging pytest<8.0 pytest-cov pytest-cases
```

On Windows it generates an error: `The system cannot find the file specified.`

The relevant section of the build log is:
```
  + pip install packaging 'pytest<8.0' pytest-cov pytest-cases
  The system cannot find the file specified.
Error: Command ['pip', 'install', 'packaging', 'pytest<8.0', 'pytest-cov', 'pytest-cases'] failed with code 1. None
```

How should the version specification be done to work on Windows (as well as linux and mac)? I've tried adding quotes and quotes with a space between the package and version specification, neither worked.

Thanks for the package and the help!!

CI link: https://github.com/RadioAstronomySoftwareGroup/pyuvdata/actions/runs/7732465269/job/21082460054

workflow link: https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/always-slicify/.github/workflows/publish-to-test-pypi.yaml

**Comment by ymd-h** [comment, marked as answer]
Escape `<` by `^`.
```
CIBW_TEST_REQUIRES: packaging pytest^<8.0 pytest-cov pytest-cases
```
---

I also struggled this problem, and finally solved it.
https://github.com/ymd-h/cpprb/commit/b1a2bfd1d3de0dc4ca963b8c785708599bd8ca66

On Windows, cibuildwheel executes sub command through shell (aka. cmd).

https://github.com/pypa/cibuildwheel/blob/9c75ea15c2f31a77e6043b80b1b7081372319d85/cibuildwheel/util.py#L154-L162

And cmd requires to escape `<` character by `^`.

Unfortunately, this is (probably) incompatible with Linux and macOS,
so that we might have to use configuration overrides or to use different environment variables depending OS.
https://cibuildwheel.pypa.io/en/stable/options/#overrides

---

I am not sure whether this escape should be handled by cibuildwheel or user.
If cibuildwheel escape these special characters internally, we can use same setting on both Windows and non-Windows,
however, it will introduce backward incompatibility. (It will break already escaped settings.)
